tls/extensions/s2n_client_key_share.c

```diff
@@ -342,21 +339,29 @@ static int s2n_client_key_share_recv_pq_hybrid(struct s2n_connection *conn, stru
         return S2N_SUCCESS;
     }
 
-    /* Ignore KEM groups with unexpected overall total share sizes */
-    if (key_share->blob.size != kem_group->client_share_size) {
-        return S2N_SUCCESS;
-    }
+    bool is_hybrid_share_length_prefixed = 0;
+    uint16_t actual_hybrid_share_size = key_share->blob.size;
 
-    /* Ignore KEM groups with unexpected ECC share sizes */
-    uint16_t ec_share_size = 0;
-    POSIX_GUARD(s2n_stuffer_read_uint16(key_share, &ec_share_size));
-    if (ec_share_size != kem_group->curve->share_size) {
-        return S2N_SUCCESS;
+    /* The length of the hybrid key share must be one of two possible lengths. It's internal values are either length
+     * prefixed, or they are not. If actual_hybrid_share_size is not one of these two lengths, then
+     * s2n_is_tls13_hybrid_kem_length_prefixed() will return an error. */
+    POSIX_GUARD_RESULT(s2n_is_tls13_hybrid_kem_length_prefixed(actual_hybrid_share_size, kem_group, &is_hybrid_share_length_prefixed));
+
+    if (is_hybrid_share_length_prefixed) {
+        /* Ignore KEM groups with unexpected ECC share sizes */
+        uint16_t ec_share_size = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint16(key_share, &ec_share_size));
+        if (ec_share_size != kem_group->curve->share_size) {
+            return S2N_SUCCESS;
+        }
     }
```

Comment:
Doesn't this conflict / change the behavior? s2n_is_tls13_hybrid_kem_length_prefixed produces an error if the size isn't expected, so we can't ignore groups with unexpected sizes here, right?

Reply:
You're right. It's a very small behavior change, but we don't need to be so strict, and can keep the previous more lax behavior.